subproviders/subscriptions.js

```diff
@@ -0,0 +1,131 @@
+const EventEmitter = require('events').EventEmitter
+const FilterSubprovider = require('./filters.js')
+const from = require('../util/rpc-hex-encoding.js')
+const inherits = require('util').inherits
+const utils = require('ethereumjs-util')
+
+function SubscriptionSubprovider(opts) {
+  const self = this
+
+  opts = opts || {}
+
+  EventEmitter.apply(this, Array.prototype.slice.call(arguments))
+  FilterSubprovider.apply(this, [opts])
+
+  this.subscriptions = {}
+}
+
+inherits(SubscriptionSubprovider, FilterSubprovider)
+
+// a cheap crack at multiple inheritance
+// I don't really care if `instanceof EventEmitter` passes...
+Object.assign(SubscriptionSubprovider.prototype, EventEmitter.prototype)
+
+// preserve our constructor, though
+SubscriptionSubprovider.prototype.constructor = SubscriptionSubprovider
+
+SubscriptionSubprovider.prototype.subscribe = function(payload, cb) {
+  const self = this
+  let method = () => {}
+
+  switch (payload.params[ 0 ]) {
+    case 'logs':
+      let options = payload.params[ 1 ]
+
+      method = self.newLogFilter.bind(self, options)
+      break
+    case 'newPendingTransactions':
+      method = self.newPendingTransactionFilter
+      break
+    case 'newHeads':
+      method = self.newBlockFilter
+      cb(null, id)
+      return
+    case 'syncing':
+    default:
+      cb(new Error('unsupported subscription type'))
+      return
+  }
+
+  method(function(err, id) {
+    if (!err) {
+      self.subscriptions[id] = payload.params[0]
+      self.filters[id].on('data', function(results) {
+        if (!Array.isArray(results)) {
+          result = [results]
+        }
+
+        var notificationHandler = self._notificationHandler.bind(self, id, payload.params[0])
+        results.forEach(notificationHandler)
+        self.filters[id].clearChanges()
+      })
+    }
+    cb(err, id)
+  })
+}
+
+SubscriptionSubprovider.prototype._notificationHandler = function (id, subscriptionType, result) {
+  const self = this
+  if (subscriptionType === 'newHeads') {
+    result = {
```

Comment:
should break this off as a utility method

Reply:
defo